### ISSUE TITLE
BigGAN/StyleGAN: make conditional training work

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#10]
BigGAN/StyleGAN work well for unconditional inputs, and for categorical-classified inputs, but for tags or text embeddings, we do not have working GANs. Arfa has established with StyleGAN experiments that `doc2vec` embeddings of Danbooru faces (_n_>300k), while encoding the relevant information, yield very poor generalization and StyleGAN seems to learn only a few basic dimensions like hair color despite the rich tag information. He experimented with one-hot encoding of the variables used to construct the cartoonfaces dataset, where the variables are known by construction, and found that despite the excellent visual quality and simple-as-possible one-hot-encoding of the variables, StyleGAN drops modes and levels of variables.

This is consistent with the StackGAN papers which note feeding in text embeddings of descriptive sentences resulted in StackGAN memorizing or not learning well, which they ascribe to the high-dimensional inputs being extremely unique for each image and extremely sparse. Their solution was to essentially add Gaussian noise (in a way respecting the correlation of each element of the embedding, since they are not independent), which they called 'noise augmentation' and explain as 'smoothing' the embeddings by jittering them and making the embedding more of a cloud than a single exact high-dimensional point.

Since StackGAN is the only relatively recent and successful text-to-image GAN I know of, there are not a lot of other suggestions as to what to try. We have a ton of data, but the problem is still occurring. I suggest noise augmentation.

A simple way to implement noise augmentation without messing around with multivariate gaussians would be to take a nonparametric approach: the correlation matrix is surely represented by *other* datapoints' embedding as well. So to 'noise augment' a given embedding, grab a random embedding from the dataset, and do a weighted-average of the current embedding and the random embedding.

**Comment by aiXander** [comment]
I think part of the problem here lies in the way the conditioning mechanism is incorporated in the loss. For StyleGAN2, the Discriminator outputs a tensor of shape [batch_size, cond_labels] which is then multiplied by the actual input_labels, masking out the logits for the conditioning information not present in the input (and thus destroying a lot of useful output that could be incorporated in the backward pass). Additionally, this architectural choice also confounds two rather different tasks (discriminating Real/Fake vs recovering the conditioning information).

StyleGAN2 conditioning mechanism (training/networks_styelgan2.py):
```
with tf.variable_scope('Output'):
        x = apply_bias_act(dense_layer(x, fmaps=max(labels_in.shape[1], 1)))
        if labels_in.shape[1] > 0:
            x = tf.reduce_sum(x * labels_in, axis=1, keepdims=True)
   scores_out = x
```
I think the easiest thing to try is to split off two separate heads from the Discriminator:
- One head simply predicts the usual D(x) / D(G(x)) Real/Fake scores
- Another head (which might be split off after some intermediate feature representation in the Discriminator) simply reconstructs the full conditioning input (without masking) and uses a simple cross-entropy loss that is added to the regular GAN loss.

This also allows you to play with their relative weighting and trade-off how much the conditioning mechanism is allowed to affect the realism of the samples.

Finally, it would be interesting to experiment with sampling out-of-distribution conditioning vectors (eg combining two conditioning labels together for which there is no actual training data) to see if the Generator can do zero-shot generalization. (Currently StyleGAN samples a conditioning vector randomly from the training set for each generated fake)

If the labels code for somewhat disentangled features, the D's classifier might correctly detect a superposition of conditioning labels and send the correct gradient signal to the Generator. This could be very interesting for generating novel classes of data for which we have no training cases (~CycleGAN, ...)

**Comment by gwern** [comment]
Another example of noise augmentation being necessary, it seems: ["textStyleGAN: Conditional Image Generation and Manipulation for User-Specified Content"](https://arxiv.org/pdf/2005.04909.pdf#page=3), Stap et al 2020:

![**Conditioning Augmentation** (pg3)](https://user-images.githubusercontent.com/352559/82743243-a4fc1480-9d36-11ea-8983-91d39dcd21d7.png)
